website/blog/2025/04-17-Leaner-Clusters-Lower-Bills.md

```diff
@@ -0,0 +1,315 @@
+    *   **Estimate Pod-Driven Overhead:** Using the predicted pod density for a candidate machine type (based on its capacity and the workload profile), we multiply this density by the measured `ratio_1_used_reserved_core_to_pods` and `ratio_1_used_reserved_gi_to_pods` to estimate the required `kube-reserved` CPU and memory, respectively. This tailors the reservation to the candidate's specific capacity and performance characteristics.
+
+4.  **Apply Thresholds for Stability:** To prevent minor fluctuations in calculated recommendations from causing constant configuration changes (increasing `kube-reserved` can trigger pod evictions), we apply thresholds (hysteresis).
+
+This tailored, data-driven approach to `kube-reserved` provides better cost optimization and enhanced stability compared to generic, workload-naive formulas.
+
+*Note on `system-reserved`:* You might wonder why we only discussed `kube-reserved` and not `system-reserved`. Similar to our reasoning against resource limits, configuring `system-reserved` can lead to unexpected CPU throttling or OOM kills for critical system processes outside Kubernetes' direct management. Therefore, Gardener focuses on configuring `kube-reserved` and relies on the kubelet's eviction mechanisms to manage overall node pressure. See also [Reserve Compute Resources for System Daemons](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#general-guidelines).
+
+## Looking Ahead: Continuous Improvement and Future Optimizations
+
+Cost optimization is an ongoing process, not a one-time fix. We're actively exploring further improvements:
+
+*   **Addressing Load Imbalances:** VPA assigns the same request to all pods in a managed group (Deployment/StatefulSet/DaemonSet). This is inefficient for workloads with inherent imbalances (e.g., controller leaders vs. followers, etcd leader vs. followers, uneven load distribution across DaemonSet pods).
+    *   **Request-Based Load Balancing:** For components like the `kube-apiserver`, default connection-based load balancing can lead to uneven load distribution that VPA handles poorly (resulting in over-provisioning for some pods, under-provisioning for others). We have implemented request-based load balancing to distribute load more evenly, allowing VPA to set more accurate requests (see [related work](https://github.com/gardener/gardener/pull/11085)).
+    *   **In-place pod resource updates** (a Kubernetes enhancement) would be particularly beneficial in the future, allowing VPA to adjust resources without requiring pod restarts, further improving efficiency and stability.
+*   **Exploring Cilium:** Initial tests suggest switching the CNI from Calico to Cilium could yield 5-10% CPU savings on worker nodes, partly because Cilium can replace kube-proxy, reducing overhead. Memory usage appears similar. Gardener supports Cilium since years.
```

Comment:
```suggestion
*   **Exploring Cilium:** Initial tests suggest switching the CNI from Calico to Cilium could yield 5-10% CPU savings on worker nodes, partly because Cilium can replace kube-proxy, reducing overhead. Memory usage appears similar and Gardener has supported Cilium for years.
```

Reply:
Yeah, reads better. Thank you.